REGRESSION(266850@main): imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window.html is a consistent failure.

cdumez opened the pull request:
#### 8572ca9d273c14ebb16289eca2c8ed80af9edb5b
<pre>
REGRESSION(266850@main): imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window.html is a consistent failure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260103">https://bugs.webkit.org/show_bug.cgi?id=260103</a>
rdar://114112210

Reviewed by Sihui Liu.

We have 2 caches for cookies, a cache for the cookie string inside Document and
a fuller in-memory cookie storage in WebCookieCache.

The new cookie store API was ignoring those caches and document.cookie could
thus return outdated cookies from those caches. To address the issue, when a
cookie gets added/removed via the cookie-store API, we now clear the cached
cookie string on the Document. We also add set the cookie in the
WebCookieCache, like it done when setting a cookie via the Document.cookie
API.

This addresses the flakiness.

* LayoutTests/TestExpectations:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::set):
* Source/WebCore/dom/Document.h:
(WebCore::Document::isDOMCookieCacheValid const):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setCookieFromDOMAsync):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::setCookieFromDOM):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::setCookieAsync const):

Canonical link: <a href="https://commits.webkit.org/269503@main">https://commits.webkit.org/269503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/32f12628e8d8792ce95751830f37c617eb8d0c42

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22727 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/398 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23814 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24634 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21038 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22986 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1472 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23255 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21963 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22967 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19720 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25487 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26817 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20839 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24663 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/274 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18118 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/213 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5423 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/341 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/276 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (approved):
r=me. Do we need to clear cache or remove entry from cache when the cookie setting fails (i.e. SetCookieFromDOMAsync returns false)?

szewai commented on Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
```diff
@@ -132,4 +168,9 @@ NetworkStorageSession& WebCookieCache::inMemoryStorageSession()
 }
 #endif
 
+bool WebCookieCache::cacheContainsStaleCookies() const
```
Maybe something like `cacheWillBeUpdatedSoon` Or `cacheMayBeOutOfSync`? `cacheContainsStaleCookies` sound like "some entries will be removed from it".

cdumez replied:
I like `cacheMayBeOutOfSync()`, will update.

szewai reviewed (approved):
r=me

cdumez committed:
REGRESSION(266850@main): imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window.html is a consistent failure.
https://bugs.webkit.org/show_bug.cgi?id=260103
rdar://114112210

Reviewed by Sihui Liu.

We have 2 caches for cookies, a cache for the cookie string inside Document and
a fuller in-memory cookie storage in WebCookieCache.

The new cookie store API was ignoring those caches and document.cookie could
thus return outdated cookies from those caches. To address the issue, when a
cookie gets added/removed via the cookie-store API, we now clear the cached
cookie string on the Document. We also add set the cookie in the
WebCookieCache, like it done when setting a cookie via the Document.cookie
API.

This addresses the flakiness.

* LayoutTests/TestExpectations:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::set):
* Source/WebCore/dom/Document.h:
(WebCore::Document::isDOMCookieCacheValid const):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setCookieFromDOMAsync):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::setCookieFromDOM):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::setCookieAsync const):

Canonical link: https://commits.webkit.org/269503@main
